activities.el

```diff
@@ -503,6 +524,45 @@ To be called from `kill-emacs-hook'."
 
 ;;;; Functions
 
+(defun activities--mapcar-window-state-leafs (state func)
+  "Return a list of leaf node values from window-state STATE.
+The returned list contains the values obtained by calling FUNC on
+each of the leaf nodes in STATE."
+  (let (values)
+    (cl-labels ((map-leafs (state func)
+		  (pcase state
+		    (`(leaf . ,_attrs)
+		     (push (funcall func state) values))
+		    ((pred proper-list-p)
+		     (if-let ((leaf-pos (cl-position 'leaf state)))
+			 (push (funcall func (cl-subseq state leaf-pos)) values)
+		       (dolist (s state) (map-leafs s func)))))))
+      (map-leafs state func))
+    (nreverse values)))
+
+(defun activities--buffers-and-files (state)
+  "Return a list of buffers and files from STATE.
+STATE is a window-state.  The returned list contains elements of
+form (BUFFER . FILE) associated with the activity."
+  (activities--mapcar-window-state-leafs
+   (activities-activity-state-window-state state)
+   (lambda (leaf)
+     (let ((buffer-rec (map-nested-elt (cdr leaf)
+				       '(parameters activities-buffer))))
+       (cons (activities-buffer-name buffer-rec)
+	     (activities-buffer-filename buffer-rec))))))
+
+(defun activities--buffers-and-files-differ-p (bf1 bf2)
+  "Return non-nil if BF1 and BF2 are not the same set of files or buffers.
+Each of BF1 and BF2 is a list of buffer and files, as returned
+from `activities--buffers-and-files'."
+  (cl-labels ((file-or-buffer (cell)
+		"Given a CELL, return the true filename or buffer.
+The CELL is a (BUFFER . FILE) cons.  If the file is nil, BUFFER is returned."
+		(if (cdr cell) (file-truename (cdr cell)) (car cell))))
+    (not (seq-set-equal-p (mapcar #'file-or-buffer bf1)
+			  (mapcar #'file-or-buffer bf2)))))
```

Comment:
Very nitpicky, but I'd prefer to name these arguments, e.g. `bf-a` and `bf-b`, because they aren't sequential in nature.  (Maybe just an idiosyncrasy of how my mind works.)